### PR TITLE
avoid suppressing exception in `FitLoop`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Resolves a small TODO in the FitLoop.
Not all samplers have a `set_epoch` method, hence we were suppressing an AttributeError. However, suppressing exceptions is not great because we don't know where they might come from. Instead, we should either catch and deal with the exception, or, as done here, replace it with a conditional check.

part of #7938 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGMT !

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -181,8 +180,7 @@ def on_advance_start(self) -> None:
             self.trainer.train_dataloader.load_state_dict(self._dataloader_state_dict)
             self._dataloader_state_dict = {}
 
-        # TODO: specify the possible exception
-        with suppress(Exception):
+        if callable(getattr(self.trainer.train_dataloader.sampler, "set_epoch", None)):
```
Comment (ananthsub) [inline pytorch_lightning/loops/fit_loop.py:183]: does this assume `sampler` is set? our current dataloader API allows for train_dataloader to be an iterable: https://github.com/PyTorchLightning/pytorch-lightning/blob/ff7305f74d4e097a45b6a4d8c0fff6d4f5aaa386/pytorch_lightning/plugins/training_type/training_type_plugin.py#L204

fyi @ninginthecloud

Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:183]: @ananthsub in PyTorch DataLoaders it's not possible to set no sampler, there will always be a default one depending on the what the value of shuffle is. 

Lightning does not support any iterables, only dataloaders and lists or dicts of dataloaders. The type information is: 

```python
TRAIN_DATALOADERS = Union[
    DataLoader,
    Sequence[DataLoader],
    Sequence[Sequence[DataLoader]],
    Sequence[Dict[str, DataLoader]],
    Dict[str, DataLoader],
    Dict[str, Dict[str, DataLoader]],
    Dict[str, Sequence[DataLoader]],
]
```

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:183]: @ananthsub could you or one of your team @daniellepintz @yifuwang @four4fish  open a feature request with your requirements and expectations for adding support for non-DataLoader iterables?
